Use headless ("Agg") matplotlib backend in CI

kousu opened the pull request:
<!-- Hi, and thank you for submitting a Pull Request! The checklist below is a brief summary of steps found in the NeuroPoly Contributing Guidelines, which can be found here: https://www.neuro.polymtl.ca/software/contributing. 
-->

## Checklist

#### GitHub

- [x] I've given this PR a concise, self-descriptive, and meaningful title
- [x] I've linked relevant issues in the PR body
- [x] I've applied [the relevant labels](https://www.neuro.polymtl.ca/software/contributing#pr_labels) to this PR
- [x] I've assigned a reviewer

<!-- For the title, please observe the following rules:
	- Provide a concise and self-descriptive title
	- Do not include the applicable issue number in the title, do it in the PR body
	- If the PR is not ready for review, convert it to a draft.
-->

#### PR contents

- [ ] I've consulted [ivadomed's internal developer documentation](https://github.com/ivadomed/ivadomed/wiki) to ensure my contribution is in line with any relevant design decisions
- [ ] I've added [relevant tests](https://github.com/ivadomed/ivadomed/wiki/tests) for my contribution
- [ ] I've updated the [relevant documentation](https://github.com/ivadomed/ivadomed/wiki/documentation) for my changes, including argparse descriptions, docstrings, and ReadTheDocs tutorial pages

## Description

Force Agg matplotlib backend to avoid problems in CI. CI should be headless anyway, there's no need to try the other six backends.

## Linked issues

Fixes #1132, which is the most recent symptom, but it's not the only tricky thing that can come up when mixing and matching rendering backends and headless test machines.

Commit:
Update run_tests.yml

Force Agg matplotlib backend

Fixes #1132

kousu commented:
We use the same solution in SCT, but there we set it as a pytest setting:

https://github.com/spinalcordtoolbox/spinalcordtoolbox/blob/20d2f599c6bccffbcc0a852ad6acd3b71d509457/testing/conftest.py#L28-L33

@joshuacwnewton and I had a debate about which file was better to set this variable in: https://github.com/spinalcordtoolbox/spinalcordtoolbox/pull/3420#discussion_r652790703

kousu commented:
Drat, it didn't work:

![Screenshot 2022-05-17 at 20-20-54 Use headless ( Agg ) matplotlib backend by kousu · Pull Request #1133 · ivadomed_ivadomed](https://user-images.githubusercontent.com/987487/168932661-4dcd0992-f651-4d8a-9ae8-f868db0eb242.png)

I don't get it, that should have worked.

joshuacwnewton commented on .github/workflows/run_tests.yml:
```diff
@@ -19,6 +19,11 @@ env:
+  # bug-prone in the constrained environment of CI
+  # Tip from a matplotlib dev: https://github.com/spinalcordtoolbox/spinalcordtoolbox/issues/3388#issuecomment-846091012
+  # Ref: https://matplotlib.org/stable/users/explain/backends.html
+  MPLBACKEND: 'Agg'
```
Could this be relevant, possibly? 

https://github.com/ivadomed/ivadomed/blob/bd904e5e139fa6a437abfc225216f6057824f1e3/ivadomed/utils.py#L225

kousu replied:
Thank you for your eyes! When I got to that point I was just tired of the problem, so I'm really happy you picked up my baton to keep us going.

It could be either. `use()` _sets a **global**_, so its leaking into other tests, [e.g.](https://github.com/ivadomed/ivadomed/runs/6480501528?check_suite_focus=true):

```
ivadomed/metrics.py:305: in plot_roc_curve
    plt.figure()
[...]
../../../hostedtoolcache/Python/3.8.12/x64/lib/python3.8/tkinter/__init__.py:2272: in __init__
    self._loadtk()
```

```
testing/unit_tests/test_mixup.py:27: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
ivadomed/mixup.py:36: in mixup
    save_mixup_sample(ofolder, data, targets, lambda_tensor)
ivadomed/mixup.py:63: in save_mixup_sample
    plt.figure(figsize=(20, 10))
[...]
../../../hostedtoolcache/Python/3.8.12/x64/lib/python3.8/tkinter/__init__.py:2272: in __init__
    self._loadtk()
```

```
 testing/functional_tests/test_visualize_transforms.py:21: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
ivadomed/scripts/visualize_transforms.py:183: in main
    run_visualization(input=args.input, config=args.config, number=int(args.number),
ivadomed/scripts/visualize_transforms.py:171: in run_visualization
    imed_utils.plot_transformed_sample(before,
ivadomed/utils.py:210: in plot_transformed_sample
    plt.figure(figsize=(20, 10))
[...]
../../../hostedtoolcache/Python/3.8.12/x64/lib/python3.8/tkinter/__init__.py:2272: in __init__
    self._loadtk()
```

None of those three `figure()` calls are near the `use()` call but it's still up in their business.

kousu replied:
Nice! Those were indeed the final pieces. CI is green: https://github.com/ivadomed/ivadomed/actions/runs/2344760419

![Screenshot 2022-05-18 at 06-47-36 Use headless ( Agg ) matplotlib backend · ivadomed_ivadomed@23e018f](https://user-images.githubusercontent.com/987487/169021844-da46e3eb-ab1c-4c65-92f9-a04ceaead117.png)

Commit:
Drop internal matplotlib.use() calls.

Leave the choice of backend to the user/environment;
trying to guess causes more trouble than it saves IMO.

Co-authored-by: Joshua Newton <joshuacwnewton@gmail.com>

kousu commented on ivadomed/scripts/visualize_and_compare_testing_models.py:
```diff
@@ -21,7 +21,7 @@
 matplotlib.rcParams['toolbar'] = 'None'  # Remove buttons
 
-gui_env = ['TKAgg', 'GTKAgg', 'Qt4Agg', 'WXAgg']
+gui_env = ['Agg']
```
This is just experimental. I'll probably revert this.

But I don't think this loop should be here at all. matplotlib already does this loop: https://matplotlib.org/stable/users/explain/backends.html

> Without a backend explicitly set, Matplotlib automatically detects a usable backend based on what is available on your system and on whether a GUI event loop is already running. The first usable backend in the following list is selected: MacOSX, QtAgg, GTK4Agg, Gtk3Agg, TkAgg, WxAgg, Agg. The last, Agg, is a non-interactive backend that can only write to files.

kousu replied:
So, I'm pretty sure this whole loop is redundant and can be replaced with just the last line of the block! Because matplotlib already does it. Check this:

```
[kousu@nigiri ~]$ python -c 'import matplotlib; print(matplotlib.get_backend())'
QtAgg
[kousu@nigiri ~]$ DISPLAY="" python -c 'import matplotlib; print(matplotlib.get_backend())'
agg
```

i.e. when I tell it I have no GUI, it automatically falls back to agg.

kousu commented on ivadomed/scripts/visualize_and_compare_testing_models.py:
```diff
@@ -21,21 +21,7 @@
-# If none works
-if selected_gui_env == []:
-    from matplotlib import pyplot as plt
-
+if matplotlib.get_backend() == "agg":
```
I'm pretty sure this loop is redundant: https://matplotlib.org/stable/users/explain/backends.html

> Without a backend explicitly set, Matplotlib automatically detects a usable backend based on what is available on your system and on whether a GUI event loop is already running. The first usable backend in the following list is selected: MacOSX, QtAgg, GTK4Agg, Gtk3Agg, TkAgg, WxAgg, Agg.

and calling `use()` in the middle of tests is messing the tests up by leaking state from one test to another, because `use()` is setting an internal global variable and hence causing some tests to use TkAgg or QtAgg or just agg whatever might be chosen depending on the order the tests happen to run in.

kousu commented:
@joshuacwnewton helped me make it work!

kanishk16 reviewed (approved):
Apart from just importing pyplot in `visualize_and_compare_testing_models.py`, LGTM!

kousu commented:
> Apart from just importing pyplot in `visualize_and_compare_testing_models.py`, LGTM!

Do you want

```
from matplotlib import pyplot as plt
```

to be moved to the top of the file instead of removed?

The tests passed without that, do you think it's really necessary?

joshuacwnewton commented:
> The tests passed without that, do you think it's really necessary?

The `visualize_and_compare_testing_models.py` is only [~16% covered](https://coveralls.io/builds/49246718/source?filename=ivadomed%2Fscripts%2Fvisualize_and_compare_testing_models.py) (though, for some reason Coveralls isn't giving line-by-line coverage). So, I'd be skeptical of trusting the test suite here?

Is it possible that the `onclick` and `visualize_and_compare_models` functions aren't tested at all?

kanishk16 commented:
> > Apart from just importing pyplot in `visualize_and_compare_testing_models.py`, LGTM!
> 
> Do you want
> 
> ```
> from matplotlib import pyplot as plt
> ```
> 
> to be moved to the top of the file instead of removed?

Yup
> The tests passed without that, do you think it's really necessary?

I guess that'd be necessary if someone runs it as a script

https://github.com/ivadomed/ivadomed/blob/bd904e5e139fa6a437abfc225216f6057824f1e3/ivadomed/scripts/visualize_and_compare_testing_models.py#L111-L115

Moreover, I believe there are no tests yet for it.

kousu commented:
> Moreover, I believe there are no tests yet for it.


Oh true, I didn't notice that. I'll put it in then merge :)

Commit:
Update visualize_and_compare_testing_models.py
